### PR TITLE
Change PIN feature

### PR DESCRIPTION
**Pull request by michellewong793**
### Description

Create new Change PIN feature for user. 

Adds analytics to change pin feature according to https://docs.google.com/spreadsheets/d/1rnSj4fjCNE1Z0Qxjn_cpBe_AFKQgS4RlKo2NLuYAahs/edit#gid=0

Adds tests for the feature 

### Tested

Tested on Android so far
### Backwards compatibility

Backwards compatible!

### CODE REVIEW

File: packages/mobile/src/navigator/Navigator.tsx [hunk 1]
```diff
@@ -192,9 +192,13 @@ const pincodeSetScreenOptions = ({
...
+  const changePin = route.params?.changePin
   const title = isVerifying
     ? i18n.t('onboarding:pincodeSet.verify')
+    : changePin
+    ? i18n.t('onboarding:pincodeSet.changePIN')
     : i18n.t('onboarding:pincodeSet.create')
+
   return {
     ...nuxNavigationOptions,
     headerTitle: () => (
```
Comment (michellewong793) [inline packages/mobile/src/navigator/Navigator.tsx:204]: still need to update this with an empty subtitle but having trouble with using an empty string here

File: packages/mobile/locales/en-US/onboarding.json [hunk 2]
```diff
@@ -29,7 +29,7 @@
   "selectCountryCode": "Select Country Code",
   "pincodeSet": {
     "create": "Create a PIN",
-    "verify": "Enter PIN again",
+    "changePIN": "Change PIN",
```
Comment (gnardini) [inline packages/mobile/locales/en-US/onboarding.json:32]: Is this change on purpose? The `verify` is used in the onboarding flow, we should probably not delete it :)

File: packages/mobile/locales/es-419/accountScreen10.json [hunk 3]
```diff
@@ -8,6 +8,7 @@
   "removeAccountDetails": "Al reinicializar se borrará tu cuenta de este dispositivo. Tu cuenta mantendrá los fondos que tenga, pero solo la podrás recuperar con tu Clave de Cuenta.",
   "legal": "Legal",
   "securityAndData": "Seguridad y datos",
+  "changePIN": "Cambiar PIN",
```
Comment (gnardini) [inline packages/mobile/locales/es-419/accountScreen10.json:11]: We shouldn't add translations for strings, that's handled by Crowdin now

File: packages/mobile/src/account/Settings.tsx [hunk 4]
```diff
@@ -320,6 +320,25 @@ export class Account extends React.Component<Props, State> {
...
+      .then((pinIsCorrect) => {
+        if (pinIsCorrect) {
+          this.props.navigation.navigate(Screens.PincodeSet, {
+            isVerifying: false,
+            changePin: true,
+          })
+        }
+      })
+      .catch((error) => {
+        Logger.error(`${'NavigationService'}@onPress`, 'PIN ensure error', error)
```
Comment (gnardini) [inline packages/mobile/src/account/Settings.tsx:336]: nit: We don't need the `${}` if we're just writing a string. We could have something like:
`Logger.error(`NavigationService@onPress`, 'PIN ensure error', error)`

File: packages/mobile/src/account/Settings.tsx [hunk 5]
```diff
@@ -320,6 +320,25 @@ export class Account extends React.Component<Props, State> {
...
+      })
+
+    // on success, navigate to PincodeSet (with prop for change Pin design)
```
Comment (gnardini) [inline packages/mobile/src/account/Settings.tsx:339]: nit: This comment is a bit disconnected from the other, maybe put the two together at the top?

File: packages/mobile/src/navigator/Navigator.tsx [hunk 6]
```diff
@@ -191,15 +191,22 @@ const pincodeSetScreenOptions = ({
 }: {
   route: RouteProp<StackParamList, Screens.PincodeSet>
 }) => {
-  const isVerifying = route.params?.isVerifying
-  const title = isVerifying
-    ? i18n.t('onboarding:pincodeSet.verify')
+  const changePin = route.params?.changePin
```
Comment (gnardini) [inline packages/mobile/src/navigator/Navigator.tsx:194]: This is losing the `isVerifying` part. We could simplify by doing:
```
const isVerifying = route.params?.isVerifying
const changePin = route.params?.changePin
let title
if (isVerifying) title = i18n.t('onboarding:pincodeSet.verify')
else if (changePin) title = i18n.t('onboarding:pincodeSet.changePIN')
else title = i18n.t('onboarding:pincodeSet.create')
```
(this assumes that we want the title to be the verifying one the second time).

another solution is welcome too of course, but lets make sure we take into account the `verifying` thing.

File: packages/mobile/src/navigator/Navigator.tsx [hunk 7]
```diff
@@ -191,15 +191,22 @@ const pincodeSetScreenOptions = ({
...
+  const title = changePin
+    ? i18n.t('onboarding:pincodeSet.changePIN')
     : i18n.t('onboarding:pincodeSet.create')
+
   return {
     ...nuxNavigationOptions,
-    headerTitle: () => (
-      <HeaderTitleWithSubtitle title={title} subTitle={i18n.t('onboarding:step', { step: '2' })} />
-    ),
+    headerTitle: () =>
```
Comment (gnardini) [inline packages/mobile/src/navigator/Navigator.tsx:201]: This is doing duplicate checks and looks a bit weird :) . Can we just leave the second option and put a space in the conditional? Something like:
```
<HeaderTitleWithSubtitle
          title={title}
          subTitle={changePin ? ' ' : i18n.t('onboarding:step', { step: '2' })}
        />
```

File: packages/mobile/src/navigator/types.tsx [hunk 8]
```diff
@@ -161,10 +161,10 @@ export type StackParamList = {
   }
   [Screens.PincodeEnter]: {
     withVerification?: boolean
-    onSuccess: (pin: string) => void
-    onCancel: () => void
+    onSuccess?: (pin: string) => void
+    onCancel?: () => void
   }
-  [Screens.PincodeSet]: { isVerifying: boolean } | undefined
```
Comment (gnardini) [inline packages/mobile/src/navigator/types.tsx:167]: Did you remove the `| undefined` on purpose?

File: packages/mobile/src/pincode/PincodeEnter.tsx [hunk 9]
```diff
@@ -88,12 +88,11 @@ class PincodeEnter extends React.Component<Props, State> {
 
   render() {
     const { t } = this.props
-    const { pin, errorText } = this.state
```
Comment (gnardini) [inline packages/mobile/src/pincode/PincodeEnter.tsx:91]: Any reason we deleted this?

File: packages/mobile/src/pincode/PincodeSet.tsx [hunk 10]
```diff
@@ -61,8 +62,11 @@ export class PincodeSet extends React.Component<Props, State> {
   }
 
   navigateToNextScreen = () => {
-    if (this.props.choseToRestoreAccount) {
+    if (this.props.route.params.changePin) {
+      navigate(Screens.Settings)
+    } else if (this.props.choseToRestoreAccount) {
       navigate(Screens.ImportWallet)
+      // if this prop is passed, then navigate back to Settings with a toast
```
Comment (gnardini) [inline packages/mobile/src/pincode/PincodeSet.tsx:69]: This comment looks out of place here and there's no toast being fired here as well

File: packages/mobile/src/pincode/PincodeSet.tsx [hunk 11]
```diff
@@ -84,6 +88,7 @@ export class PincodeSet extends React.Component<Props, State> {
     return this.state.pin1 === pin
   }
 
+  // Function for PIN to be changed
```
Comment (gnardini) [inline packages/mobile/src/pincode/PincodeSet.tsx:91]: This is not true, this function is also used during onboarding. Maybe we can just delete it?

File: packages/mobile/src/pincode/PincodeSet.tsx [hunk 12]
```diff
@@ -104,6 +109,7 @@ export class PincodeSet extends React.Component<Props, State> {
       this.props.setPincode(PincodeType.CustomPin)
       ValoraAnalytics.track(OnboardingEvents.pin_set)
       this.navigateToNextScreen()
+      Logger.showMessage('PIN changed')
```
Comment (gnardini) [inline packages/mobile/src/pincode/PincodeSet.tsx:112]: During onboarding we don't want this toast to show up. Also, it should be translated :)

File: packages/mobile/src/pincode/PincodeSet.tsx [hunk 13]
```diff
@@ -118,22 +124,24 @@ export class PincodeSet extends React.Component<Props, State> {
...
     const { pin1, pin2, errorText } = this.state
 
     return (
-      <SafeAreaView style={styles.container}>
+      <SafeAreaView style={changePin ? styles.changePinContainer : styles.container}>
         <DevSkipButton onSkip={this.navigateToNextScreen} />
         {isVerifying ? (
-          // Verify
           <Pincode
+            title="Enter PIN again to confirm"
```
Comment (gnardini) [inline packages/mobile/src/pincode/PincodeSet.tsx:136]: These texts should be translated. Also, this is changing the onboarding flow, do we want them to show up always or only if `changePin` is true?

**Review by gnardini — COMMENTED** [review]
Looks really good, great job! Left mostly some style related comments and I know you've been working on adding tests as well :)

File: packages/mobile/locales/en-US/onboarding.json (hunk 2 above)
Comment (michellewong793) [reply packages/mobile/locales/en-US/onboarding.json:32]: oops! took it out thinking i had added it on accident-  thanks!

File: packages/mobile/src/navigator/Navigator.tsx (hunk 6 above)
Comment (michellewong793) [reply packages/mobile/src/navigator/Navigator.tsx:194]: i took it out because in the designs, there actually isn't a difference between isVerifying and not... the title is always just Change PIN! figma here: 
image

File: packages/mobile/src/pincode/PincodeSet.tsx (hunk 10 above)
Comment (michellewong793) [reply packages/mobile/src/pincode/PincodeSet.tsx:69]: thanks! this is one of the ones i wrote way earlier and forgot about. thanks!

File: packages/mobile/src/pincode/PincodeSet.tsx (hunk 11 above)
Comment (michellewong793) [reply packages/mobile/src/pincode/PincodeSet.tsx:90]: deleted the comment

File: packages/mobile/src/pincode/PincodeSet.tsx (hunk 12 above)
Comment (michellewong793) [reply packages/mobile/src/pincode/PincodeSet.tsx:111]: thanks! added conditional

File: packages/mobile/src/pincode/PincodeSet.tsx (hunk 13 above)
Comment (michellewong793) [reply packages/mobile/src/pincode/PincodeSet.tsx:135]: i think that the text is shared, according to this figma. is this the only change you're talking about, the title? 
image

Comment (michellewong793) [reply packages/mobile/src/pincode/PincodeSet.tsx:135]: https://www.figma.com/file/JHrjTQVHlAPnNTzcoEGeVy/Wallet-4.0?node-id=11117%3A2318 figma link here

File: packages/mobile/locales/es-419/accountScreen10.json (hunk 3 above)
Comment (michellewong793) [reply packages/mobile/locales/es-419/accountScreen10.json:11]: oh ok! how does Crowdin handle translations? is there anything i have to add for that?

Comment (gnardini) [reply packages/mobile/locales/es-419/accountScreen10.json:11]: No need to add anything, it happens automatically. I don't know the details too well, but whenever a new string is pushed to `main` Crowdin picks it up and asks in its UI (which is used by the product team afaik) for translations. Whenever we want to sync what's on Crowdin with the code we can tell Crowdin to create a PR with all the translation changes.

File: packages/mobile/src/navigator/Navigator.tsx (hunk 6 above)
Comment (gnardini) [reply packages/mobile/src/navigator/Navigator.tsx:194]: You're right! Seems like we got rid of the `isVerifying` thing.

File: packages/mobile/locales/es-419/accountScreen10.json (hunk 3 above)
Comment (gnardini) [reply packages/mobile/locales/es-419/accountScreen10.json:11]: We do need to delete the `es` and `pt` translations though because we might get conflicts when Crowding opens its PR otherwise

Comment (michellewong793) [reply packages/mobile/locales/es-419/accountScreen10.json:11]: oh nice!! thanks :)

File: packages/mobile/src/account/Settings.test.tsx [hunk 14]
```diff
@@ -60,4 +70,33 @@ describe('Account', () => {
...
   })
+
+  it('navigates to PincodeSet screen if entered PIN is correct', async () => {
+    const tree = render(
+      <Provider store={createMockStore({})}>
+        <Settings {...getMockStackScreenProps(Screens.Settings)} />
+      </Provider>
+    )
+    fireEvent.press(tree.getByTestId('ChangePIN'))
+    mockedNavigate.mockImplementationOnce((_, params) => {
```
Comment (gnardini) [inline packages/mobile/src/account/Settings.test.tsx:81]: This test (and the next) is not really doing what we want them to.

We're writing the mock implementation of the `navigate` function _after_ pressing the button. We should do that before. It will run automatically when the button is pressed. After pressing the button we should only write `expect`s to verify that what we expected to happen happened.

Since we're mocking the `ensurePincode` we just need to provide the logic we want `ensurePincode` to have. In this test, we want it to succeed by returning a promise that has `true` as its argument (the `true` is the `pinIsCorrect` variable).

So, it should be something like:

```
const tree = (...)
mockedEnsurePincode.mockImplementationOnce(() => Promise.resolve(true))
fireEvent.press(tree.getByTestId('ChangePIN'))
expect(navigate).toHaveBeenCalledWith(arguments)
```

For the next test, instead of returning a promise that resolves with `true` we should throw an error, which will cause the `ensurePincode` call to go the the `catch` callback.

File: packages/mobile/src/pincode/PincodeSet.tsx [hunk 15]
```diff
@@ -104,6 +107,9 @@ export class PincodeSet extends React.Component<Props, State> {
       this.props.setPincode(PincodeType.CustomPin)
       ValoraAnalytics.track(OnboardingEvents.pin_set)
       this.navigateToNextScreen()
+      if (this.props.route.params.changePin) {
+        Logger.showMessage('PIN changed')
```
Comment (gnardini) [inline packages/mobile/src/pincode/PincodeSet.tsx:111]: This should still be translated :)

File: packages/mobile/src/pincode/PincodeSet.tsx (hunk 13 above)
Comment (gnardini) [reply packages/mobile/src/pincode/PincodeSet.tsx:136]: Ok, sounds good, the `verifying` text seems to be shared but the initial one isn't (`Create a new PIN` vs `Change your PIN` or something like that).

Also, needs to be on a locales file to be translated

File: packages/mobile/src/account/Settings.tsx (hunk 4 above)
Comment (michellewong793) [reply packages/mobile/src/account/Settings.tsx]: sure!

File: packages/mobile/src/pincode/PincodeSet.tsx (hunk 13 above)
Comment (michellewong793) [reply packages/mobile/src/pincode/PincodeSet.tsx:136]: oh i see, ill put a conditional looks like originally it's just empty, so will just check that changePin prop is true to display this title.

How do the locales files work? Which file should i add this title to?

File: packages/mobile/src/navigator/Navigator.tsx (hunk 7 above)
Comment (michellewong793) [reply packages/mobile/src/navigator/Navigator.tsx]: agreed, changed

File: packages/mobile/src/navigator/types.tsx (hunk 8 above)
Comment (michellewong793) [reply packages/mobile/src/navigator/types.tsx:167]: unintentional

Comment (michellewong793) [reply packages/mobile/src/navigator/types.tsx:167]: will add it back! was trying to look for where it is used though, any ideas? :o

File: packages/mobile/src/pincode/PincodeSet.tsx (hunk 15 above)
Comment (michellewong793) [reply packages/mobile/src/pincode/PincodeSet.tsx:111]: using the i18n library?

Comment (michellewong793) [reply packages/mobile/src/pincode/PincodeSet.tsx]: waiting for the right translations from the slack channels :3

File: packages/mobile/src/account/Settings.test.tsx (hunk 14 above)
Comment (michellewong793) [reply packages/mobile/src/account/Settings.test.tsx:81]: ooooooo thank you gonza! this makes a lot of sense, will rewrite

File: packages/mobile/src/pincode/PincodeSet.tsx (hunk 13 above)
Comment (michellewong793) [reply packages/mobile/src/pincode/PincodeSet.tsx]: lmk if that worked, i changed the string to be the translated version

File: packages/mobile/src/pincode/PincodeSet.tsx (hunk 15 above)
Comment (gnardini) [reply packages/mobile/src/pincode/PincodeSet.tsx]: No need to translate manually, sorry my comment wasn't clear 😅  We just need to put the string in the locale file in English and the product team will take care of translating to other languages with Crowdin

File: packages/mobile/src/account/Settings.test.tsx [hunk 16]
```diff
@@ -90,4 +99,33 @@ describe('Account', () => {
...
     expect(tree).toMatchSnapshot()
   })
+
+  it('navigates to PincodeSet screen if entered PIN is correct', async () => {
+    const tree = render(
+      <Provider store={createMockStore({})}>
+        <Settings {...getMockStackScreenProps(Screens.Settings)} />
+      </Provider>
+    )
+    mockedNavigate.mockImplementationOnce(() => {
```
Comment (gnardini) [inline packages/mobile/src/account/Settings.test.tsx:109]: Do these tests pass?

The function that we should mock that resolves to a promise with true is `ensurePincode`, not `navigate`. We don't need to provide any behaviour to `navigate`, we need to add this logic to `ensurePincode`.

Also, you're not returning the promise, so this implementation is not returning anything :/

File: packages/mobile/src/account/Settings.test.tsx [hunk 17]
```diff
@@ -90,4 +99,33 @@ describe('Account', () => {
...
+  })
+
+  it('does not navigate to PincodeSet screen if entered PIN is incorrect', async () => {
+    const tree = render(
+      <Provider store={createMockStore({})}>
+        <Settings {...getMockStackScreenProps(Screens.Settings)} />
+      </Provider>
+    )
+    mockedNavigate.mockImplementationOnce((error) => {
+      return error
```
Comment (gnardini) [inline packages/mobile/src/account/Settings.test.tsx:126]: The `navigate` doesn't receive an `error` param, we should mock `ensurePincode` instead of `navigate` and have this throw an error or return `Promise.reject()`

File: packages/mobile/src/pincode/PincodeSet.test.tsx [hunk 18]
```diff
@@ -134,4 +136,36 @@ describe('Pincode', () => {
     await flushMicrotasksQueue()
     expect(getByText('pincodeSet.pinsDontMatch')).toBeDefined()
   })
+
+  it('navigates to the Settings screen after successfully changing PIN', async () => {
+    const { getByTestId, rerender } = render(
+      <Provider store={mockStore}>
+        <PincodeSet {...mockScreenProps} />
```
Comment (gnardini) [inline packages/mobile/src/pincode/PincodeSet.test.tsx:143]: `mockScreenProps` has `isVerifying` as true, so it'll start with that. And then we're rerendering with `isVerifying` as `false`? Did you run this test and does it pass? Things seem to be in the incorrect order.

File: packages/mobile/src/pincode/PincodeSet.tsx [hunk 19]
```diff
@@ -125,22 +132,24 @@ export class PincodeSet extends React.Component<Props, State> {
...
+            title={i18n.t('onboarding:pincodeSet.verify')}
             errorText={errorText}
             pin={pin2}
             onChangePin={this.onChangePin2}
             onCompletePin={this.onCompletePin2}
           />
         ) : (
-          // Create
           <Pincode
+            title={changePin ? 'Create a new PIN' : ' '}
```
Comment (gnardini) [inline packages/mobile/src/pincode/PincodeSet.tsx:152]: This `Create a new PIN` text should be on a file in the `locales` folder as well

**Review by gnardini — COMMENTED** [review]
Getting closer :)

File: packages/mobile/locales/es-419/accountScreen10.json [hunk 20]
```diff
@@ -8,6 +8,8 @@
...
```
[the remark this answers is not in the document]
Comment (michellewong793) [reply packages/mobile/locales/es-419/accountScreen10.json]: thanks! kept the english versions

File: packages/mobile/src/pincode/PincodeSet.tsx (hunk 19 above)
Comment (michellewong793) [reply packages/mobile/src/pincode/PincodeSet.tsx]: got it!

File: packages/mobile/src/pincode/PincodeSet.test.tsx (hunk 18 above)
Comment (michellewong793) [reply packages/mobile/src/pincode/PincodeSet.test.tsx:143]: after the user has entered their new pin, `isVerifying` should be true, so yeah i think it should be true

File: packages/mobile/src/account/Settings.test.tsx (hunk 17 above)
Comment (michellewong793) [reply packages/mobile/src/account/Settings.test.tsx]: ah i see, does the updated code make sense?

File: packages/mobile/src/account/Settings.test.tsx (hunk 16 above)
Comment (michellewong793) [reply packages/mobile/src/account/Settings.test.tsx]: been having trouble running it D: i changed the implementation of the mock for `ensurePincode`. does that look correct?

File: packages/mobile/src/account/Settings.tsx [hunk 21]
```diff
@@ -329,6 +329,24 @@ export class Account extends React.Component<Props, State> {
...
+            isVerifying: false,
+            changePin: true,
+          })
+        }
+      })
+      .catch((error) => {
+        ValoraAnalytics.track(SettingsEvents.change_pin_current_pin_error)
+        Logger.error('NavigationService@onPress', 'PIN ensure error', error)
+      })
+  }
```
Comment (tarikbellamine) [inline packages/mobile/src/account/Settings.tsx:348]: Nit: Can you please use the async/await with try/catch block pattern? I'm not sure if that's our official team preference but that's what we use most throughout the codebase

File: packages/mobile/src/account/Settings.tsx [hunk 22]
```diff
@@ -329,6 +329,24 @@ export class Account extends React.Component<Props, State> {
     this.setState({ showRevokeModal: false })
   }
 
+  goToChangePin = () => {
+    ValoraAnalytics.track(SettingsEvents.change_pin_start)
+    ensurePincode()
+      .then((pinIsCorrect) => {
+        if (pinIsCorrect) {
+          ValoraAnalytics.track(SettingsEvents.change_pin_current_pin_entered)
+          this.props.navigation.navigate(Screens.PincodeSet, {
```
Comment (tarikbellamine) [inline packages/mobile/src/account/Settings.tsx:338]: Nit: Can you please use the `navigate` function from `NavigationService` instead of calling it from props? I know that's not how the rest of the file has it but that's because this file is old and in need of a refactor :)

The benefit is that the function has built in type checking + logging so it's preferable.

File: packages/mobile/src/pincode/PincodeSet.tsx [hunk 23]
```diff
@@ -94,8 +98,14 @@ export class PincodeSet extends React.Component<Props, State> {
   onCompletePin1 = () => {
     if (this.isPin1Valid(this.state.pin1)) {
       this.props.navigation.setParams({ isVerifying: true })
+      if (this.props.route.params.changePin) {
```
Comment (tarikbellamine) [inline packages/mobile/src/pincode/PincodeSet.tsx:101]: I think you need optional chaining on all of these references to `changePin` because `this.props.route.params` can be undefined

File: packages/mobile/src/pincode/PincodeSet.test.tsx [hunk 24]
```diff
@@ -134,4 +136,36 @@ describe('Pincode', () => {
...
+        />
+      </Provider>
+    )
+
+    // Verify pin
+    mockPin.split('').forEach((number) => fireEvent.press(getByTestId(`digit${number}`)))
+    jest.runAllTimers()
+    await flushMicrotasksQueue()
+
+    expect(navigateClearingStack).toBeCalledWith(Screens.Settings)
```
Comment (tarikbellamine) [inline packages/mobile/src/pincode/PincodeSet.test.tsx:169]: I think you want to expect `navigate` here not `navigateClearingStack`

File: packages/mobile/src/pincode/PincodeSet.test.tsx [hunk 25]
```diff
@@ -134,4 +136,36 @@ describe('Pincode', () => {
...
+      </Provider>
+    )
+
+    // Create pin
+    mockPin.split('').forEach((number) => fireEvent.press(getByTestId(`digit${number}`)))
+    jest.runAllTimers()
+    await flushMicrotasksQueue()
+    expect(mockScreenProps.navigation.setParams).toBeCalledWith({
+      isVerifying: true,
+      changePin: true,
```
Comment (tarikbellamine) [inline packages/mobile/src/pincode/PincodeSet.test.tsx:153]: `setParams` is never called with `changePin` so you'll want to remove this line

File: packages/mobile/src/pincode/PincodeSet.test.tsx (hunk 18 above)
Comment (tarikbellamine) [reply packages/mobile/src/pincode/PincodeSet.test.tsx:143]: I think what you want here is for `mockScreenProps` to have to have `changePin: true` so that the `PincodeSet` screen knows that we are navigating here from the `Settings` screen. `isVerifying` should only be set on the 2nd render because that's how the screen knows that this is the 2nd time the user is entering their pin and it should be checking if it's the same as the first entry

**Review by tarikbellamine — COMMENTED** [review]
Nice work! Almost there :)

File: packages/mobile/src/pincode/PincodeSet.tsx (hunk 23 above)
Comment (michellewong793) [reply packages/mobile/src/pincode/PincodeSet.tsx:101]: thanks!

File: packages/mobile/src/account/Settings.tsx (hunk 22 above)
Comment (michellewong793) [reply packages/mobile/src/account/Settings.tsx:338]: sure! sounds good

File: packages/mobile/src/pincode/PincodeSet.test.tsx (hunk 25 above)
Comment (michellewong793) [reply packages/mobile/src/pincode/PincodeSet.test.tsx]: ah i see i might have just been confused that this had to be set again

File: packages/mobile/src/pincode/PincodeSet.test.tsx (hunk 24 above)
Comment (michellewong793) [reply packages/mobile/src/pincode/PincodeSet.test.tsx]: i was thinking this too but for some reason, when i try to call navigate here it wont show any screens being called, and with navigateClearingStack the VerificationEducationScreen is actually popping up

Comment (michellewong793) [reply packages/mobile/src/pincode/PincodeSet.test.tsx]: image

Comment (michellewong793) [reply packages/mobile/src/pincode/PincodeSet.test.tsx]: image

File: packages/mobile/src/account/Settings.test.tsx [hunk 26]
```diff
@@ -90,4 +99,29 @@ describe('Account', () => {
...
     expect(tree).toMatchSnapshot()
   })
+
+  it('navigates to PincodeSet screen if entered PIN is correct', async () => {
+    const tree = render(
+      <Provider store={createMockStore({})}>
+        <Settings {...getMockStackScreenProps(Screens.Settings)} />
+      </Provider>
+    )
+    mockedEnsurePincode.mockImplementation(() => true)
```
Comment (gnardini) [inline packages/mobile/src/account/Settings.test.tsx:109]: `ensurePincode` doesn't return a boolean, it returns a `Promise<boolean>`, so this test is probably not going to pass :/ You can run individual tests locally using `jest src/account/Settings.test.tsx` or `yarn test src/account/Settings.test.tsx`.

To fix this, instead of `true` here you can return `Promise.resolve(true)` and the same thing for the other test case below. Also, since you have to wait for a promise to resolve it's likely that you'll need to add `jest.runAllTimers()` after firing the press button event, which makes sure that anything async is run.

**Review by gnardini — COMMENTED** [review]
The tests are getting closer :) but I don't see where the actual changing of the PINs is being done. Have you checked that this changes your PIN and you can then use the new one and not the old one?

An easy way to check is to try to open the account key screen to see if the new one is accepted.

File: packages/mobile/src/pincode/authentication.ts [hunk 27]
```diff
@@ -252,6 +253,30 @@ export async function checkPin(pin: string, account: string) {
   return hashForPin === correctHash
 }
 
+export async function updatePin(account: string, oldPin: string, newPin: string) {
+  try {
+    clearPasswordCaches()
```
Comment (tarikbellamine) [inline packages/mobile/src/pincode/authentication.ts:258]: Would it be better to do this only when `updated` is successful?

File: packages/mobile/src/pincode/PincodeSet.tsx [hunk 28]
```diff
@@ -111,7 +137,20 @@ export class PincodeSet extends React.Component<Props, State> {
       this.props.setPincode(PincodeType.CustomPin)
       ValoraAnalytics.track(OnboardingEvents.pin_set)
       this.navigateToNextScreen()
+      if (this.isChangingPin()) {
+        const updated = await updatePin(this.props.account, this.state.oldPin, pin2)
```
Comment (tarikbellamine) [inline packages/mobile/src/pincode/PincodeSet.tsx:141]: Can you just create an `oldPin` variable here using `getCachedPin(DEFAULT_CACHE_ACCOUNT) ?? '' })` and pass it in directly instead of storing to state?

**Review by tarikbellamine — COMMENTED** [review]
Looks good to me but I'm reluctant to approve. It's a sensitive part of the code base and I don't know it very well so I think @jeanregisser should take a look.

File: packages/mobile/ios/Podfile.lock [hunk 29]
```diff
@@ -4,7 +4,7 @@ PODS:
   - Adjust/Core (4.26.1)
   - Analytics (4.1.3)
   - boost-for-react-native (1.63.0)
-  - CeloBlockchain (0.0.336)
+  - CeloBlockchain (0.0.351)
```
[the remark this answers is not in the document]
Comment (gnardini) [reply packages/mobile/ios/Podfile.lock:7]: It was updated to an intermediate one, not the latest. Not sure if this is the best PR to update but since I was already updating `react-native-geth` and I had to play around with this version a bit I decided to leave the upgrade

File: packages/mobile/src/pincode/PincodeSet.tsx (hunk 28 above)
Comment (gnardini) [reply packages/mobile/src/pincode/PincodeSet.tsx:141]: The reason not to do that is that the `getCachedPin` might expire in the time between the user enters their PIN the first time and they enter the new one. The cache lasts for 5 minutes, so it's unlikely to happen, but if it did the user would be asked for their old password again after inputting the new one, which will likely be pretty confusing.

I'll add a comment to the code since it's not very clear.

Comment (gnardini) [reply packages/mobile/src/pincode/PincodeSet.tsx:141]: Thinking it over, if a user takes more than 5 minutes to input their new PIN maybe we want to expire their request or something and have them start over? For security reasons?

Comment (tarikbellamine) [reply packages/mobile/src/pincode/PincodeSet.tsx:142]: Oh I see. I don't really think there's much of a security concern...probably not worth the effort for an unlikely edge case.

Another option is to make a property on the component that holds this value. Whenever I see something in component state I expect that it will be changed and trigger re-renders so it was confusing for me to see it there. In a functional component we'd define this with `const` instead of `useState` right?

Comment (gnardini) [reply packages/mobile/src/pincode/PincodeSet.tsx:142]: In a functional component I'd probably use a `useMemo` hook instead. Using just a const would re-fetch from the cache in every render, but the useMemo would cache the result. Not sure if there's an easy equivalent for class components :/

**Review by jeanregisser — COMMENTED** [review]
Nice!!

File: packages/mobile/src/pincode/PincodeSet.tsx [hunk 30]
```diff
@@ -111,7 +138,20 @@ export class PincodeSet extends React.Component<Props, State> {
       this.props.setPincode(PincodeType.CustomPin)
       ValoraAnalytics.track(OnboardingEvents.pin_set)
       this.navigateToNextScreen()
+      if (this.isChangingPin()) {
```
[the remark this answers is not in the document]
Comment (gnardini) [reply packages/mobile/src/pincode/PincodeSet.tsx:139]: Moved to before :)

**Comment by Celoqa** [comment]
https://valora.testquality.com defect [D4](https://valora.testquality.com/project/12128/result/2397005/defect/7770). Do not remove or edit.

File: packages/mobile/src/pincode/PincodeSet.tsx [hunk 31]
```diff
@@ -109,9 +136,23 @@ export class PincodeSet extends React.Component<Props, State> {
     if (this.isPin1Valid(pin1) && this.isPin2Valid(pin2)) {
       setCachedPin(DEFAULT_CACHE_ACCOUNT, pin1)
```
Comment (jeanregisser) [inline packages/mobile/src/pincode/PincodeSet.tsx:137]: For the `changePin` case, it seems we should only call this when `updatePin` is successful, right?

**Review by jeanregisser — COMMENTED** [review]
One last question 👀

File: packages/mobile/src/pincode/PincodeSet.tsx (hunk 31 above)
Comment (gnardini) [reply packages/mobile/src/pincode/PincodeSet.tsx]: Yeah, good catch, fixed!

File: packages/mobile/src/pincode/PincodeSet.tsx [hunk 32]
```diff
@@ -107,11 +134,25 @@ export class PincodeSet extends React.Component<Props, State> {
...
     const { pin1 } = this.state
     if (this.isPin1Valid(pin1) && this.isPin2Valid(pin2)) {
-      setCachedPin(DEFAULT_CACHE_ACCOUNT, pin1)
       this.props.setPincode(PincodeType.CustomPin)
-      ValoraAnalytics.track(OnboardingEvents.pin_set)
+      if (this.isChangingPin()) {
+        const updated = await updatePin(this.props.account, this.state.oldPin, pin2)
+        if (updated) {
+          ValoraAnalytics.track(SettingsEvents.change_pin_new_pin_confirmed)
+          Logger.showMessage(i18n.t('accountScreen10:pinChanged'))
```
Comment (jeanregisser) [inline packages/mobile/src/pincode/PincodeSet.tsx:142]: It seems we should also `setCachedPin(DEFAULT_CACHE_ACCOUNT, pin1)` when updated, right?
otherwise the old pin would remain in cache.

Comment (gnardini) [reply packages/mobile/src/pincode/PincodeSet.tsx:142]: The `updatePin` function manages the cache, it clears it and saves the new pin there.

**Review by jeanregisser — APPROVED** [review]
Ship it!! ✅ 🎉

**Comment by Celoqa** [comment]
Hi @MuckT @jeanregisser  We have verified this task on latest Android internal build V1.16.1 & Testflight build V1.16.1 (67) and updated the status in below sheet;
**Sheet link:** https://docs.google.com/spreadsheets/d/1wYi5kbe_NUXuhc9xtcbT9dtyrVbPXqNBrAETfSx0R58/edit#gid=2064669594
CC: @gnardini @michellewong793
